File: apollo-api/src/main/java/com/apollographql/apollo/api/ResponseWriter.java
package com.apollographql.apollo.api;

import java.util.List;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface ResponseWriter {
  void writeString(@NotNull ResponseField field, @Nullable String value);

  void writeInt(@NotNull ResponseField field, @Nullable Integer value);

  void writeLong(@NotNull ResponseField field, @Nullable Long value);

  void writeDouble(@NotNull ResponseField field, @Nullable Double value);

  void writeBoolean(@NotNull ResponseField field, @Nullable Boolean value);

  void writeCustom(@NotNull ResponseField.CustomTypeField field, @Nullable Object value);

  void writeObject(@NotNull ResponseField field, @Nullable ResponseFieldMarshaller marshaller);

  <T> void writeList(@NotNull ResponseField field, @Nullable List<T> values, @NotNull ListWriter<T> listWriter);

  interface ListWriter<T> {
    void write(@Nullable List<T> items, @NotNull ListItemWriter listItemWriter);
  }

  interface ListItemWriter {
    void writeString(@Nullable String value);

    void writeInt(@Nullable Integer value);

    void writeLong(@Nullable Long value);

    void writeDouble(@Nullable Double value);

    void writeBoolean(@Nullable Boolean value);

    void writeCustom(@NotNull ScalarType scalarType, @Nullable Object value);

    void writeObject(@Nullable ResponseFieldMarshaller marshaller);

    <T> void writeList(@Nullable List<T> items, @NotNull ListWriter<T> listWriter);
  }
}

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/internal/ResponseWriter.kt
package com.apollographql.apollo.api.internal

import com.apollographql.apollo.api.ResponseField
import com.apollographql.apollo.api.CustomScalar
import com.apollographql.apollo.api.Response

interface ResponseWriter {
  fun writeString(field: ResponseField, value: String?)

  fun writeInt(field: ResponseField, value: Int?)

  fun writeLong(field: ResponseField, value: Long?)

  fun writeDouble(field: ResponseField, value: Double?)

  fun writeBoolean(field: ResponseField, value: Boolean?)

  fun writeCustom(field: ResponseField, value: Any?)

  fun writeObject(field: ResponseField, block: ((ResponseWriter) -> Unit)?)

  fun <T: Any> writeList(field: ResponseField, values: List<T?>?, block: (item: T, listItemWriter: ListItemWriter) -> Unit)

  interface ListWriter<T> {
    fun write(items: List<T>?, listItemWriter: ListItemWriter)
  }

  interface ListItemWriter {
    fun writeString(value: String)

    fun writeInt(value: Int)

    fun writeLong(value: Long)

    fun writeDouble(value: Double)

    fun writeBoolean(value: Boolean)

    fun writeCustom(customScalar: CustomScalar, value: Any)

    fun writeObject(block: ((ResponseWriter) ->Unit))

    fun <T: Any> writeList(items: List<T?>, block: (item: T, listItemWriter: ListItemWriter) -> Unit)
  }
}
